[Bug] [Mater] The process always runs when the process contains subprocess and params deliver

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.


### What happened

The process always runs when the process contains subprocess and params deliver

Process have subprocessess
![image](https://github.com/apache/dolphinscheduler/assets/20885366/ef44dcf1-7e15-48e0-bed1-a75663aace6f)

Have params deliver situation
![image](https://github.com/apache/dolphinscheduler/assets/20885366/65ba2cbe-119b-4d1a-9160-e407d0bbf49b)


When running this process and process will still run 

Master throw exception 

```
2023-09-05 03:01:32.862 +0000 [ERROR] org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable:[428] - [WorkflowInstance-750084][TaskInstance-930027] - Task finish failed, get a exception, will remove this taskInstance from completeTaskSet
java.lang.NullPointerException: null
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.setVarPoolValue(WorkflowExecuteRunnable.java:1271)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.getPreVarPool(WorkflowExecuteRunnable.java:1234)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.submitStandByTask(WorkflowExecuteRunnable.java:1854)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.submitPostNode(WorkflowExecuteRunnable.java:1372)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.taskFinished(WorkflowExecuteRunnable.java:413)
	at org.apache.dolphinscheduler.server.master.event.TaskStateEventHandler.handleStateEvent(TaskStateEventHandler.java:68)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.handleEvents(WorkflowExecuteRunnable.java:293)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
2023-09-05 03:01:32.862 +0000 [ERROR] org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable:[308] - [WorkflowInstance-750084][TaskInstance-930027] - State event handle error, get a unknown exception, will retry this event: TaskStateEvent(processInstanceId=750084, taskInstanceId=930027, taskCode=0, status=null, type=TASK_STATE_CHANGE, key=750086-0-750084-930027, channel=null, context=null)
java.lang.NullPointerException: null
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.setVarPoolValue(WorkflowExecuteRunnable.java:1271)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.getPreVarPool(WorkflowExecuteRunnable.java:1234)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.submitStandByTask(WorkflowExecuteRunnable.java:1854)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.submitPostNode(WorkflowExecuteRunnable.java:1372)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.taskFinished(WorkflowExecuteRunnable.java:413)
	at org.apache.dolphinscheduler.server.master.event.TaskStateEventHandler.handleStateEvent(TaskStateEventHandler.java:68)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.handleEvents(WorkflowExecuteRunnable.java:293)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)

```


### What you expected to happen

Running as expect

### How to reproduce

Needs two or more masters, only one master can not reproduce this bug. 

The reason is when a subprocess, will dispatch to different masters, And when the task(sub process) running is finished. Masters notify logic has an issue, not updating the task's end-time field will cause this bug.

### Anything else

_No response_

### Version

3.1.x

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/event/TaskStateEventHandler.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/event/TaskStateEventHandler.java
@@ -90,6 +90,7 @@ public class TaskStateEventHandler implements StateEventHandler {
             if (iTaskProcessor.taskInstance().getState().isFinished()) {
                 if (iTaskProcessor.taskInstance().getState() != task.getState()) {
                     task.setState(iTaskProcessor.taskInstance().getState());
+                    task.setEndTime(iTaskProcessor.taskInstance().getEndTime());
                 }
                 workflowExecuteRunnable.taskFinished(task);
             }

